crates/rust-analyzer/src/config.rs

```diff
@@ -234,13 +235,21 @@ impl Config {
                 }
                 // otherwise configure command customizations
                 _ => {
-                    if let Some(FlycheckConfig::CargoCommand { command, extra_args, all_targets, all_features })
+                    if let Some(FlycheckConfig::CargoCommand { command, extra_args, all_targets, all_features, features })
                         = &mut self.check
                     {
                         set(value, "/checkOnSave/extraArgs", extra_args);
                         set(value, "/checkOnSave/command", command);
                         set(value, "/checkOnSave/allTargets", all_targets);
-                        set(value, "/checkOnSave/allFeatures", all_features);
+                        if let Some(new_all_features) = get(value, "/checkOnSave/allFeatures") {
+                            *all_features = new_all_features;
+                        } else {
+                            *all_features = self.cargo.all_features;
+                        }
+                        set(value, "/checkOnSave/features", features);
```

Comment:
I think here we also should check for `null`, rather than is empty. I imagine the case where you have non-empty set for `cargo metadata`, but want explicitelly to override it with `[]` for `checkOnSave`.